output/cloud/expv2/output.go

```diff
@@ -133,6 +133,16 @@ func (o *Output) Start() error {
 			TLSConfig: insights.ClientTLSConfig{
 				Insecure: false,
 			},
+			RetryConfig: insights.RetryConfig{
+				RetryableStatusCodes: "UNAVAILABLE,INTERNAL",
+				MaxAttempts:          5,
+				PerRetryTimeout:      1 * time.Second,
+				BackoffConfig: insights.BackoffConfig{
+					Enabled:        true,
+					JitterFraction: 0.1,
+					WaitBetween:    1 * time.Second,
+				},
+			},
```

Comment:
We need to adjust this config. Currently, we send data using the insights client on its own goroutine, but the calls themselves are *blocking*. The [flushRequestMetadatas()](https://github.com/grafana/k6/blob/1df72a0a1b23d7b0827813f6b709c7ad7bb4193b/output/cloud/expv2/output.go#L279-L291) method is invoked periodically every [TracesPushInterval](https://github.com/grafana/k6/blob/1df72a0a1b23d7b0827813f6b709c7ad7bb4193b/output/cloud/expv2/output.go#L147). What's more, this interval is also used as the context deadline [here](https://github.com/grafana/k6/blob/1df72a0a1b23d7b0827813f6b709c7ad7bb4193b/output/cloud/expv2/output.go#L282). This means that we currently get just 1 second *total* to send a request using this client. Therefore, configuring `MaxAttemps` to `5` with `PerRetryTimeout` equal to `1s` makes little sense.

I see a few paths forward:
1. We make the [requestMetadataFlusher](https://github.com/grafana/k6/blob/1df72a0a1b23d7b0827813f6b709c7ad7bb4193b/output/cloud/expv2/flush.go#L154) non-blocking and increase the context deadline required to flush traces so that it encompasses what we defined in the retry config.
2. We keep the requestMetadataFlusher blocking, increasing the context deadline overall. This may be dangerous as we could theoretically block for a long time while, at the same time, the traces accumulate in memory. This means that after some unavailability, the next request might be huge (in terms of the number of metadatas to send) compared to the mean.

I'd like to hear @codebien opinion on this as well.

Reply:
Thanks for the feedback you are totally right. If we decide to implement 1, I understand that we should create a channel to receive if the execution was successful to handle the error properly [here](https://github.com/grafana/k6/blob/1df72a0a1b23d7b0827813f6b709c7ad7bb4193b/output/cloud/expv2/output.go#L271), right?

There would be another alternative that I don't know makes sense that would be to keep the retries within the [TracesPushInterval](https://github.com/grafana/k6/blob/1df72a0a1b23d7b0827813f6b709c7ad7bb4193b/output/cloud/expv2/output.go#L147) of 1 second. I would like to hear the thoughts of @codebien as well.

Reply:
> If we decide to implement 1, I understand that we should create a channel to receive if the execution was successful to handle the error properly [here](https://github.com/grafana/k6/blob/1df72a0a1b23d7b0827813f6b709c7ad7bb4193b/output/cloud/expv2/output.go#L271), right?

We don't need more sophisticated error handling at this point. We don't return any sophisticated errors from the backend at this point. In the case of (1), I thought more about "fire and forget". Maybe in the future we add more logic to it.

Reply:
Let's follow the pattern you already established. It is easy to implement.

So, we would have to:
1. Add default timeout to `insights.Client`. It can be 1 minute as well. It will make retries very robust.
2. Add `TracesPushConcurrency` option to `cloudapi.Config`
3. Call `requestMetadatasFlusher.flush(ctx)` with `context.Background()`.
4. Introduce similar logic to `runFlushWorkers()`, but for our `flushRequestMetadatas()` method.

@vortegatorres WDYT?

Reply:
I agree to follow the same strategy.